### PR TITLE
@clerk/expo React compatibility improvements

### PR DESCRIPTION
**Pull request by nikosdouvlis**
## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

## Description

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.

EventTarget is not yet supported in most non-browser environments (eg: expo). Instead of poly filling the dependency, this PR simply replaces EventTarget with a usual handler array implementation.

We've also moved react and react-dom to peerDependencies and marked both as external dependencies for the headless clerk-js builds. Our webpack config has been updated to allow per-entry point configuration for prod builds.

### CODE REVIEW

**Review by dimkl — APPROVED** [review]

File: packages/clerk-js/src/core/events.ts [hunk 1]
```diff
@@ -5,48 +5,45 @@ export const events = {
...
-    this.addEventListener(eventName, handler);
-  }
+    eventToHandlersMap.get(event)?.push(handler);
+  };
 
-  dispatch<E extends ClerkEvent, P extends EventPayload[E]>(eventName: E, data: P): void {
-    this.dispatchEvent(new CustomEvent(eventName, { detail: data }));
-  }
+  const dispatch = <E extends ClerkEvent>(event: E, payload: EventPayload[E]) => {
+    (eventToHandlersMap.get(event) || []).forEach(h => typeof h === 'function' && h(payload));
```
Comment (dimkl) [inline packages/clerk-js/src/core/events.ts:27]: ❓ Do we need `typeof h === 'function'` since we are using `EventHandler` type ?

**Review by SokratisVidros — COMMENTED** [review]
Makes sense.

File: packages/clerk-js/src/core/events.ts (hunk 1 above)
Comment (nikosdouvlis) [reply packages/clerk-js/src/core/events.ts:27]: On the type level, we don't. However, I think these little sanity checks are good practice as an `any`  type on the side of the caller (or the part of the codebase that fires the dispatch call) might crash the host app.

File: packages/clerk-js/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+      ...common({ mode }),
+      ...commonForProd(),
+      ...externalsForProd(),
+    },
+  };
+
+  const configs = [];
+
+  for (const entry of Object.keys(entryToSourceMap)) {
+    console.log('Will build for ', entry);
```
Comment (panteliselef) [inline packages/clerk-js/webpack.config.js:131]: ❓ Are we keeping this log ?

File: packages/clerk-js/webpack.config.js [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+      plugins: [
+        new ReactRefreshWebpackPlugin({ overlay: { sockHost: 'js.lclclerk.com' } }),
+        ...(env.serveAnalyzer ? [new BundleAnalyzerPlugin()] : []),
+      ],
+      devtool: 'eval-cheap-source-map',
+      entry: './src/index.browser.ts',
+    },
+    devServerOutput(),
+  );
+  console.log(config);
```
Comment (panteliselef) [inline packages/clerk-js/webpack.config.js:180]: ❓ Are we keeping this log ?

File: packages/clerk-js/webpack.config.js [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    output: {
+      path: path.resolve(__dirname, 'dist'),
+      filename: '[name].js',
+      libraryTarget: 'umd',
+      globalObject: 'globalThis',
+    },
+  };
+};
+/** @type { () => (import('webpack').Configuration) } */
+const externalsForProd = () => {
```
Comment (panteliselef) [inline packages/clerk-js/webpack.config.js:89]: 🙃 `externalsForProdHeadless` or `externalsForHeadless` ?

Comment (nikosdouvlis) [reply packages/clerk-js/webpack.config.js:89]: I like this more, changed :)

**Pull request merged**